### ISSUE TITLE
Feature Request: externals should be able to specify arbitrary expressions

### ISSUE THREAD

**Issue by fresheneesz** [webpack/webpack#5631]
**Do you want to request a *feature* or report a *bug*?**

Feature

**What is the current behavior?**

Currently, externals only allow specifying a single variable as the value to use from the global scope. Eg `{externals: {'gem/Style':'Gem.Style'}` will return a bundle that loads `root["Gem.Style"]`. This is obviously not what I expected to load here.

**What is the expected behavior?**

I'd like to be able to specify an arbitrary expression in order to find the module. For example, I'd like to be able to specify externals like  `{externals: {"gem/Style": {expression: ".Gem.Style"}}}` so that webpack would render the external as `root.Gem.Style`. This would also allow arbitrary module loading in whatever form the user needs.

**If this is a feature request, what is motivation or use case for changing the behavior?**

My motivation is that I have a module that expects modules to be loaded individually in common js, like this:

```
var Text =require("gem/Text")
var CheckBox = require("gem/CheckBox")
```

but when loaded as a UMD module, expects the user to access those same things via:

```
var Text = Gem.Text
var CheckBox = Gem.CheckBox
```

Because I'm loading modules that don't exist in a single variable, I'm currently forced to work around this by manually creating some boiler plate that translates between these things, by having a configuration like: `{externals: {"gem/Style":"Style"}}` and adding `var Style = Gem.Style` at the top of my html page. 

**Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**

Webpack v3.5.5

**Comment by sokra** [comment]
`externals: {'gem/Style':['Gem','Style']}` -> `root["Gem"]["Style"]`

see also https://webpack.js.org/configuration/externals/

**Comment by fresheneesz** [comment]
Oh I see. I guess I missed that part. Thanks! 

I will say I found the docs on externals a little hard to follow, and definitely hard to scan for what I needed. Maybe I'll come back and help improve that section at some point.

**Comment by fresheneesz** [comment]
@sokra Hmm, I'm getting an error when trying your suggestion:

```
WebpackOptionsValidationError: Invalid configuration object. Webpack has been initialised using a co
nfiguration object that does not match the API schema.
 - configuration.externals should be one of these:
   string | object { <key>: string | object | boolean } | function | RegExp | [(recursive)]
   Specify dependencies that shouldn't be resolved by webpack, but should become dependencies of the
 resulting bundle. The kind of the dependency depends on `output.libraryTarget`.
   Details:
    * configuration.externals should be a string.
    * configuration.externals['gem/Style'] should be a string.
    * configuration.externals['gem/Style'] should be an object.
    * configuration.externals['gem/Style'] should be a boolean.
    * configuration.externals['gem/Style'] should be one of these:
      string | object | boolean
    * configuration.externals['gem/Text'] should be a string.
    * configuration.externals['gem/Text'] should be an object.
    * configuration.externals['gem/Text'] should be a boolean.
    * configuration.externals['gem/Text'] should be one of these:
      string | object | boolean
    * configuration.externals should be an instance of function
      `function(context, request, callback(err, result))` The function is called on each dependency.
.
    * configuration.externals should be an instance of RegExp
      Every matched dependency becomes external..
    * configuration.externals should be an array:
      [string | object { <key>: string | object | boolean } | function | RegExp | [(recursive)]]
```

My externals are:
```
externals : {
   proto:'proto', gem:'Gem', "gem/Style":['Gem','Style'], "gem/Text":["Gem",'Text']
}
```

**Comment by sokra** [comment]
hmm looks like a bug in the schema. Do you want to send a PR allowing it?

**Comment by fresheneesz** [comment]
@sokra Ok I can try to that. But looking at the repository, I can't figure out the appropriate place to add a test for this case. `Validation.test` doesn't seem like the right place, cause all those tests expect errors, whereas I need to expect no error. Maybe somewhere in `configCases/externals` would be the right place, but its not clear to me. Where should I add a test for this?

**Comment by sokra** [comment]
yep config cases seems good.

**Comment by fresheneesz** [comment]
Ok, after slogging through the code and ajv, I successfully added a single test case to test the issue. I don't know enough about the breadth of `externals` cases, so I didn't add any more than that single case. I also fixed the portion that was failing validation. It looks like the actual handling of externals in this format isn't working correctly, and after about an hour of looking through the code, I'm just not sure what to do about it. 

I've committed the new test and validation fix here: https://github.com/fresheneesz/webpack/commit/5da9e868e7ffce6e2e8d252c32ab5debc5fc6178 . It looks like someone with deeper knowledge than me is needed to efficiently fix this.

**Comment by sokra** [comment]
Could you just submit a PR with the current state? We can discuss it there.

**Comment by fresheneesz** [comment]
@sokra Ok here's a pull request for it: https://github.com/webpack/webpack/pull/5640

**Comment by fresheneesz** [comment]
@sokra 

Honestly, after looking through the code and cases for this, I don't think the additional conciseness that all these different forms of `externals` allows is worth the huge cognitive load and complexity it introduces. I'd recommend deprecating those forms and just allow the following forms:

```
externals: function(module) {
   if(shouldBeExcluded(module) {
     return "root."+module // exclude, replacing with the given code string
   } else {
     return undefined // don't exclude
   }
}
```
```
externals: {
   <modulePath>: "root."+modulePath
}
```

This is far simpler, and provides as much power as the user could need. Anyone that's generating their webpack config can write helper functions to make putting together externals as concise as they want:

```
function createExternals(externals) {
   var resultExternals = {}
   for(var external in externals) {
     resultExternals[external] = "root."+externals[external]
   })
   return resultExternals
}

var webpackConfig = {
   externals: createExternals({"a":"a","b":"b","c/d":"c.d"})
}
```

Its far more important that people can quickly understand how to use webpack via the docs than it is to allow a pure webpack config to be as concise as possible.
